db/python/tables/sample.py

```diff
@@ -687,3 +693,55 @@ async def get_samples_with_missing_participants_by_internal_id(
             )
         )
         return samples
+
+    async def get_monthly_samples_count_per_project(
+        self, project_ids: list[int] | None = None
+    ) -> dict[int, dict[date, int]]:
+        """
+        Get a map of {project_id: {month:count} for list of project_ids
+        If project_ids is empty, return all projects
+        """
+
+        where_str = 'WHERE project in :project_ids' if project_ids else ''
+        values = {'project_ids': project_ids} if project_ids else {}
+
+        _query = f"""
+        WITH t AS(
+            SELECT project, id, min(row_start) as sample_first_date
+            FROM sample FOR SYSTEM_TIME ALL
+            {where_str}
+            GROUP BY project, id
+        )
+        SELECT project, CAST(EXTRACT(YEAR FROM sample_first_date) AS INTEGER) AS year, CAST(EXTRACT(MONTH FROM sample_first_date) AS INTEGER) AS month, COUNT(*) AS count
+        FROM t
+        GROUP BY project, year, month
+        ORDER BY project, year, month
+        """
+
+        rows = await self.connection.fetch_all(_query, values)
+        result: dict[int, dict[date, int]] = {}
+        accumulated_sample_count: dict[int, int] = {}
+
+        # results are sorted by project, year, month
+        # accumulate projects previous months into the processed month
+        for r in rows:
+            proj = r['project']
+            if proj not in result:
+                result[proj] = {}
+
+            # append previous months count to currently processed month
+            accumulated_sample_count[proj] = (
+                accumulated_sample_count.get(proj, 0) + r['count']
+            )
+            result[proj][date.fromisoformat(f"{r['year']}-{r['month']:02d}-01")] = (
+                accumulated_sample_count[proj]
+            )
+
+        # append all the months up to the current month
+        for proj, months in result.items():
+            current_month = date.today().replace(day=1)
+            while current_month not in months:
```

Comment:
I think this could go back in time infinitely if the months list is empty to start with?